### PR TITLE
Added LazyBatchNorm(1,2,3)d

### LINKED ISSUE #48: Feature request: LazyBatchNorm
In my project, I implemented LazyBatchNorm like this and I think pytorch-pfn-extras should have the similar feature:

```
class LazyBatchNorm1d(
    ppe.nn.modules.lazy.LazyInitializationMixin, torch.nn.BatchNorm1d
):
    """
    LazyBatchNorm1d is a lazy version of BatchNorm1d.  It does not require to
    specify the number of the input parameters beforehand.
    """

    lazy_parameter_names = ("weight", "bias")

    def __init__(self, num_features, *args, **kwargs):
        super().__init__(num_features or 0, *args, **kwargs)
        if num_features is None:
            self.num_features = None
            self.weight = ppe.nn.modules.lazy.UninitializedParameter()
            self.bias = ppe.nn.modules.lazy.UninitializedParameter()

    def forward(self, input):
        if isinstance(self.weight, ppe.nn.modules.lazy.UninitializedParameter):
            self.num_features = input.shape[-1]
            if self.affine:
                self.weight = torch.nn.Parameter(
                    self.weight.new_empty(self.num_features)
                )
                self.bias = torch.nn.Parameter(self.weight.new_empty(self.num_features))
            if self.track_running_stats:
                self.running_mean = torch.zeros(
                    self.num_features, device=self.running_mean.device
                )
                self.running_var = torch.ones(
                    self.num_features, device=self.running_var.device
                )
            self.reset_parameters()
        return super().forward(input)

    def reset_parameters(self):
        if self.lazy_parmeters_determined:
            super().reset_parameters()
```

Another discussion:
It would be good to have LazyBatchNorm instead of having LazyBatchNorm*d-s since lazy ones know its desired shapes. Any thoughts?

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #48 

Based on the implementation by @imos

### CODE REVIEW

File: pytorch_pfn_extras/nn/modules/lazy_batchnorm.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                self.running_var = torch.ones(
+                    self.num_features, device=self.running_var.device,
+                    dtype=self.running_mean.dtype
+                )
+            self.reset_parameters()
+        return super().forward(input)
+
+
+class _LazyBatchNorm(_LazyNormBase, torch.nn.modules.batchnorm._BatchNorm):
+    def forward(self, input):
```
Comment (kmaehashi) [inline pytorch_pfn_extras/nn/modules/lazy_batchnorm.py:55]: Is this override needed?

**Review by kmaehashi — CHANGES_REQUESTED** [review]

File: pytorch_pfn_extras/nn/modules/lazy_batchnorm.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import torch
+
+from pytorch_pfn_extras.nn.modules.lazy import LazyInitializationMixin
+from pytorch_pfn_extras.nn.modules.lazy import UninitializedParameter
+
+
+class _LazyNormBase(
+    LazyInitializationMixin, torch.nn.modules.batchnorm._NormBase
+):
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply pytorch_pfn_extras/nn/modules/lazy_batchnorm.py:9]: Both of them implement forward so we need to call both 😅

File: pytorch_pfn_extras/nn/modules/lazy_batchnorm.py (hunk 1 above)
Comment (emcastillo) [reply pytorch_pfn_extras/nn/modules/lazy_batchnorm.py:55]: This is how it was implemented in torch, there is a general forward to all the batchnorm implementations, and then dimension specific ones.

File: pytorch_pfn_extras/nn/modules/lazy_batchnorm.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (emcastillo) [reply pytorch_pfn_extras/nn/modules/lazy_batchnorm.py:9]: torch.nn.modules.batchnorm._NormBase holds all the batchnorm generic code, I am not overriding everything so I keep it as the base class.
I think ppe lazyconv modules are implemented like this.

**Comment by emcastillo** [comment]
PTAL

File: pytorch_pfn_extras/nn/modules/lazy_batchnorm.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                self.running_mean = torch.zeros(
+                    self.num_features, device=self.running_mean.device,
+                    dtype=self.running_mean.dtype
+                )
+                self.running_var = torch.ones(
+                    self.num_features, device=self.running_var.device,
+                    dtype=self.running_mean.dtype
+                )
+            self.reset_parameters()
+        return super().forward(input)
```
Comment (emcastillo) [inline pytorch_pfn_extras/nn/modules/lazy_batchnorm.py:52]: This calls `torch.nn.modules.batchnorm._BatchNorm.forward` so I overcomplicated it because I got a bit confused.

**Comment by kmaehashi** [comment]
LGTM!
